### ISSUE TITLE
Better Serverlist/Eliminate the serverlist gamemode duplication problem

### ISSUE THREAD

**Issue by Leystryku** [Facepunch/garrysmod-requests#1593]
When you open up your Garry's Mod and look for a server you see **~20 gamemodes on a 1080p screen**. However, **~10 of those gamemodes are RP**. And infact, basically all of them **are  actually DarkRP** but renamed to e.g. **StarwarsRP** due to having different addons.

Quite frankly, this sucks and kills gamemode variety.

I propose that perhaps we should instead have subcategories e.g. "RP" and then people could put all their "unique yet the same" gamemodes there.

That way both sides would be happy. RPers could still play on their "unique"" DarkRP rebrands and non RP-ers could have their own categories for e.g. TTT, Murder etc.

Just to avoid people flooding specific fake custom gamemodes again, the categories could even be fixed e.g. "RP, Build, " etc.

**Comment by bizzclaw** [comment]
I agree that there needs to be a better way to distinguish DarkRP servers from unique ones that are made with a lot more effort. I think a huge part of garrysmod has always been about finding some random niche server that does things in the specific way you like, so the server browser should really have features to make it easier for players to find servers, and frankly for servers to find their players.

Still, we'd have to figure out the category to put each gamemode in, and who comes up with the categories?  This could make it even more difficult to find custom servers if you're insisting all servers that are RP of any kind are all mixed together. I'm working on a RPG style PVP/PVE gamemode that, while not specifically RP because it has no rules, would still find itself forced into that category unless there was a specific "Not RP but Level Based PVP RPG" Category.

At least right now its sort of doable to tell what's DarkRP and what isn't, but I think there's a much easier and more obvious solution. The best way to eliminate the serverlist gamemode duplication problem, is to make it easier to find good servers and I think the first step towards that is giving servers the ability to communicate what they're about from the server list without player's having to join.

You've all seen it, Staff wanted! PAC3, CUSTOM!!!, people try to fill as much into the hostname as they can hoping they can catch more attention, but what can you really expect from them? As it stands that's is the single only way to say anything about your server outside of its map and gamemode, and I think this can be improved.

Even something as simple as giving server's the ability to somehow communicate a string for something like a description would go a long way.

**Comment by Leystryku** [comment]
@bizzclaw 
You thought a  lot about this.
Firstly a string to describe the server did exist and even now one existiert (gamemodename). However, people are abusing it to communicate that their gamemode is something different while it is not. (E.g. DarkRP as Gm+ addons -> StarwarsRP). Secondly, it is not possible properly doable to automatically classify the category. Already known gamemodes could be classified (e.g. some mapping for gms without the category based in gmname or a "Uncategorized" category). The server owner or gamemode creator would still be able to affect this though.

The idea is that there being fixed categories (RP,PVP, Build, etc.) Effectively prevents just anyone from introducing NEW categories (e.g. Starwarsrp  while there is RP). You would press on RP  and only then get a list of RP gamemodes (DarkRP, StarwarsRP, HogwartsRP, your gm e.g. RPG PVE). This would actually help your gamemode, especially if it's private because you would then be visible on the list without scrolling down forever.


Thirdly, while your case is mixing 2 core gameplays I'd say it still belongs in the RP or a RPG category. Lots of RPs in Gmod have PvP and Levelling - yet their "top category"/"Main Relation" is still RP. If yours is so different from RP and there's a RPG category you could shove in there. Keep in mind that the way would be "RP-> your gamemode name-> your server", so you can still clarify your full gm name (e.g. rpg lvl) there.

There's no one preventing the devs from adding enough categories. I do believe they should be kept simple though. RP, Build, Zombies, etc. so that there's no confusion.

**Comment by Leystryku** [comment]
More text:

There'd likely still be people putting their gamemode maybe in wrong categories, but the fact that these top categories would be fixed would make it pretty pointless to do so.

In the current system naming your gamemode something it is not (for example DarkRP -> HogwartsRP or DarkRP -> Harry Potter) results in you showing up on the top outside of the other RP servers.  People might think it's not DarkRP/it's a different gamemode and join because of that. This is why we have 5 of the top 7 gamemodes being DarkRP in renamed right now. Occasionally, we even have people spoofing new random gamemode names on the top 10x with some random stuff like "DarkRP remastered or whatever".
https://i.imgur.com/TEIvAs1.png

With a category/subcategory system this would be pointless because they wouldn't be able to create new categories. They'd be in a fixed category, which would:
1. decide what kind of main players they attract e.g. in RP category you'd mainly have players playing RP. By putting your RP outside of your proper category (RP) you'd essentially lose the people who're trying to play what you're hosting.
2. prevent them from creating toplevel categories, therefore we wouldn't have to worry about people flooding the main part of the serverlist with some new fake gamemodes all the time
3. promote diversity, because unlike right now where you have to scroll down like 180 gamemodes just to find your own gamemode, it'd at least be in top category it belongs to thus it wouldn't get mangled up with gamemodes with an entirely different gameplay (e.g. RP while your gamemode type  is PVP, or Build while your gamemode type  is Minigame)

**Comment by uRandomAlex** [comment]
Let me propose a list of categories and their respective gamemodes.

- **PvP**: TTT, PropHunt, Breach, Zombie Survival, Murder, BaseWars, Deathrun, DayZ
- **PvE/Campaign**: Escape from Pedobear, Half-Life Coop, Horror Story
- **Build**: Sandbox, SledBuild
- **Roleplay**: DarkRP, Star Wars RP, SCP RP, Military RP etc.
- **Minigame**: Gmod Tower, Minigame
- **Other**: Cinema, Surf, Bunny Hop, Elevator: Source

**Comment by Kefta** [comment]
Maybe a **Skill** category for Surf, Bunny Hop, and Deathrun; and a **Leisure** category for Cinema. If this is implemented, I believe the approach would be to have category view open by default that lists everything, and have an **All** category to show the full serverlist. There could be a checkbox at the top after clicking on a category to ``Open this by default`` so that people who serially play one category doesn't have to click something every time.

As for reporting the tag, Rubat suggested the gamemode.txt have a ``type`` key, and if it's unspecified, the server is put in **Other**. I'm not to sure how to enforce abuse/miscategorisation as redirect servers could be made to have a server appear in every category, so moderation might need to be strict.

**Comment by bizzclaw** [comment]
Building off the idea of using the Workshop as a reference, there could be these overarching categories, but maybe you could also add tags to your gamemode's txt file, but only read the first 3 or four, that way a server won't be able to cram them to show up in every search.

These tags could help servers communicate what features they're focussed on offering while giving players an easier time finding the server they want to play.

As for what tags could be used, these are just what I can think of off the top of my head, but a lot of the few of the ones on already the workshop happen to apply.

- **Fun**: The gamemode's sole focus is to be fun.
- **Realism**: The gamemode has an Emphasis on providing a realistic/immersive experience.
- **Build**: The gamemode has an emphasis on building.
- **Combat**: The gamemode has a focus on combat.
- **Passive**: The gamemode has a lot of passive downtimes, ideal if the server is a more serious RP style gamemode
- **Persistence**: The gamemode has a high degree of persistence, implying character and inventory systems are a focus
- **Vehicles**: The gamemode makes heavy use of Vehicles
- **NPCs**: The gamemode makes heavy use of NPCs
- **Low Rules**: The gamemode does not rely on administrative force to achieve the optimal gameplay _(You know I had to do it)_

**Comment by eldarstorm** [comment]
Have to agree. PostNukeRP was not in any form based on or uses DarkRP code, but will still get buried under the Dark RP servers

**Comment by viral32111** [comment]
> It just occurred to me that It might also be worth adding a way to specify what languages(s) your gamemode/server is intended for; that way, people can filter out servers running on languages they can't understand.

I agree with this so much, it is way better than how most server's do it these days (prefixing server name with 2 letter country code and stuff). It could just be some sort of convar sent in a2s info responses.

**Comment by Leystryku** [comment]
I do also think that the addition of language and even more would be a good idea but...

I think it'd be asking for a little too much for one github request.
I think it might be a good idea to go one by one, if this requests gets listened to, additional stuff could later be requested either way.

**Comment by NathanX-S** [comment]
paging @Kefta

**Comment by ThatLing** [comment]
> Other: Cinema, Surf, Bunny Hop, Elevator: Source

bruh, let gamemodes choose their own tags and limit it to 2

**Comment by Leystryku** [comment]
I think the refinement, like tweaking it to suits everyones specific needs can be done after this ticket/issue has been handled. There might be some people with improvement ideas or whatever - but first off, the system which could later improved shall be implemented.

**Comment by ThatLing** [comment]
Or just implement it correctly in the first place

**Comment by robotboy655** [comment]
Related issue https://github.com/Facepunch/garrysmod-requests/issues/167

**Comment by robotboy655** [comment]
The server browser was updated with many new features to hopefully help people more easily find what they are looking for, including gamemode categories. The list includes only 4 categories for the time being: `rp`, `pvp`, `pve` and `other`.

I do not feel like more categories are needed, such as "build", etc, as there's simply not that many gamemodes that would fit that or other categories. A gamemode can only have 1 category, and it works on the same rules as gamemode title and workshopID do - the category is selected based on how many servers report themselves in a category. If majority report Roleplay, that category will be assigned.

I did not create another level of categories in the gamemode list, as I feel that adds unnecessary friction for the average player, especially since on release majority of servers/gamemodes will be uncategorized. Not to mention server owners/gamemode authors updating their gamemode files to categorize their gamemodes may be extremely slow and may need further incentivizing in a future update.
